jumpscale/sals/nettools/__init__.py

```diff
@@ -32,117 +47,167 @@ def tcp_connection_test(ipaddr: str, port: int, timeout: Optional[int]):
         verify (bool, optional): boolean indication to verify the servers TLS certificate or not.
+        interval_time (int, optional): how long to wait for a response before sending a new request. Defaults to 2.
 
+    Raises:
+        ValueError: raises if not correct url
 
     Returns:
         bool: true if the test succeeds
     """
-    for _ in range(timeout):
-        try:
-            if check_url_reachable(url, timeout, verify):
-                return True
-        except:
-            pass
-
+    deadline = time.time() + timeout
+    while time.time() < deadline:
+        if check_url_reachable(url, interval_time, verify):
+            return True
+        # be gentle on system resource in case the above call to check_url_reachable() returned immediately (edge cases)
         time.sleep(1)
-    else:
-        return False
+    return False
 
 
-def check_url_reachable(url: str, timeout=5, verify=True):
+def check_url_reachable(
+    url: str, timeout: Optional[int] = 5, verify: Optional[bool] = True, fake_user_agent: Optional[bool] = True
+) -> bool:
     """Check that given url is reachable
 
     Args:
         url (str): url to test
         timeout (int, optional): timeout of test. Defaults to 5.
         verify (bool, optional): boolean indication to verify the servers TLS certificate or not.
+        fake_user_agent (bool, optional): boolean indication to fake the user-agent and act like noraml browser or not.
 
     Raises:
-        Input: raises if not correct url
+        ValueError: raises if not correct url
 
     Returns:
         bool: True if the test succeeds, False otherwise
     """
+    # fake the user-agent, to act like a normal browser
+    # because some services will block requests from python default user-agent
+    # By default urllib identifies itself as Python-urllib/x.y (e.g. Python-urllib/2.5),
+    # which may confuse the site, or just plain not work.
+    # ex: www.amazon.com, and it will looks unreachable to our code, unless we fake the user-agent.
+    HEADERS = {
+        "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_4) AppleWebKit/603.1.30 (KHTML, like Gecko) Version/10.1 Safari/603.1.30"
+    }
+    METHOD = "GET"
+
+    if timeout:
+        socket.setdefaulttimeout(timeout)
+
+    context = None
+    if not verify:
+        # opt out of certificate verification on a single connection
+        context = ssl._create_unverified_context()
+
+    req = Request(url, headers=HEADERS if fake_user_agent else {}, method=METHOD)
     try:
-        code = jumpscale.tools.http.get(url, timeout=timeout, verify=verify).status_code
-        return code == 200
-    except jumpscale.tools.http.exceptions.MissingSchema:
-        raise Input("Please specify correct url with correct scheme")
-    except jumpscale.tools.http.exceptions.ConnectionError:
+        response = urlopen(req, timeout=timeout, context=context)
+    except HTTPError:
+        # The server couldn't fulfill the request.
```

Comment:
can we log the error, using `j.logger.error("your message")` or warning according to the error level

Reply:
sounds perfect. I'll do that module wide.